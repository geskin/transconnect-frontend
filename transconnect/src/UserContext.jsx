import { createContext, useState, useEffect } from "react";

const UserContext = createContext();

export const UserProvider = ({ children }) => {
    const [currUser, setCurrUser] = useState(() => {
        const storedUser = localStorage.getItem("currUser");
        return storedUser ? JSON.parse(storedUser) : null;
    });

    // Save to localStorage whenever currUser changes
    useEffect(() => {
        if (currUser) {
            localStorage.setItem("currUser", JSON.stringify(currUser));
        } else {
            localStorage.removeItem("currUser");
        }
    }, [currUser]);

    return (
        <UserContext.Provider value={{ currUser, setCurrUser }}>
            {children}
        </UserContext.Provider>
    );
};

export default UserContext;

