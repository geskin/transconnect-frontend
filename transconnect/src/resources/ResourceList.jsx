import React, { useContext, useEffect, useState } from "react";
import UserContext from "../UserContext";
import { Link } from "react-router-dom";
import TransconnectApi from "../api";
import ResourceCard from "./ResourceCard";
import TextField from "@mui/material/TextField";
import Button from "@mui/material/Button";
import Box from "@mui/material/Box";

/** ResourceList: displays cards for every existing resource
 * 
 * auth required: none
 * 
 * unapproved resources and certain functionalities are only visible/accessible to admins
 */

const ResourceList = () => {
    const [resources, setResources] = useState([]);
    const [types, setTypes] = useState([]);
    const [searchType, setSearchType] = useState("");
    const [searchTerm, setSearchTerm] = useState("");
    const { currUser } = useContext(UserContext);

    useEffect(() => {
        const fetchResources = async () => {
            try {
                let data = await TransconnectApi.getResources();
                if (currUser.role === "USER") {
                    data = data.filter((r) => r.approved); // Only approved resources for non-admin users
                }
                setResources(data);
            } catch (err) {
                console.error("Error fetching resources", err);
            }
        };
        const fetchTypes = async () => {
            try {
                const data = await TransconnectApi.getTypes();
                let typeList = data.map((t) => t.name);
                setTypes(typeList);
            } catch (err) {
                console.error("Error fetching types", err);
            }
        };

        fetchResources();
        fetchTypes();
    }, [currUser]);

    const handleClick = async (e) => {
        e.preventDefault();
        setSearchType(e.target.value);
        // filter posts by type (can only search by one type at a time?)
        try {
            const data = await TransconnectApi.getResources(searchType);
            setResources(data);
        } catch (err) {
            console.error("Error fetching resources", err);
        }
    };

    // Handle search bar input change
    const handleChange = (e) => {
        setSearchTerm(e.target.value);
    };

    const handleSubmit = async (e) => {
        e.preventDefault();
        try {
            const data = await TransconnectApi.getResources(searchTerm);
            setResources(data);
        } catch (err) {
            console.error("Error fetching resources", err);
        }
    };

    console.debug(resources);

    return (
        <div>
            {/* Search Bar */}
            <Box sx={{ display: "flex", justifyContent: "center", alignItems: "center", gap: 1, p: 2 }}>
                <TextField
                    fullWidth
                    label="Enter search term..."
                    variant="outlined"
                    value={searchTerm}
                    onChange={handleChange}
                    sx={{ maxWidth: 600 }}
                />
                <Button type="submit" variant="contained">
                    Search
                </Button>
            </Box>


            {/* Filter Buttons */}
            <div>
                <Button value="" onClick={handleClick} variant="outlined">
                    X
                </Button>
                {types.map((t) => (
                    <Button key={t} name={t} value={t} onClick={handleClick} variant="outlined" sx={{ m: 0.5 }}>
                        {t}
                    </Button>
                ))}
            </div>

            {/* Resource Cards */}
            <div>
                {resources.length ? (
                    <div>
                        {resources.map((r) => (
                            <ResourceCard
                                key={r.id}
                                id={r.id}
                                name={r.name}
                                description={r.description}
                                url={r.url}
                                types={r.types}
                                userId={r.userId}
                                approved={r.approved}
                            />
                        ))}
                    </div>
                ) : (
                    <p className="lead white-letters">Sorry, no results were found!</p>
                )}
            </div>

            {/* Submit a Resource Button */}
            <Box sx={{ textAlign: "center", mt: 2 }}>
                <Button variant="contained" component={Link} to="/resources/new">
                    Submit a resource
                </Button>
            </Box>
        </div>
    );
};

export default ResourceList;