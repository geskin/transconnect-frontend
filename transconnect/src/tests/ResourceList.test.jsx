import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import ResourceList from "../resources/ResourceList";
import UserContext from "../UserContext";
import TransconnectApi from "../api";
import { MemoryRouter, Route, Routes } from "react-router-dom";

vi.mock("../api");
vi.mock("../resources/ResourceCard", () => ({
    __esModule: true,
    default: vi.fn(() => <div data-testid="resource-card">ResourceCard</div>),
}));

describe("ResourceList component", () => {
    const mockCurrUser = { username: "testuser", id: 1, role: "USER" };
    const mockUser = { username: "otheruser", id: 2 };
    const mockResources = [
        { id: 101, name: "Resource One", description: "First resource", approved: true, userId: 1, types: ["Type1"] },
        { id: 102, name: "Resource Two", description: "Second resource", approved: true, userId: 2, types: ["Type2"] },
    ];
    const mockTypes = ["Type1", "Type2"];

    beforeEach(() => {
        TransconnectApi.getResources.mockResolvedValue(mockResources);
        TransconnectApi.getTypes.mockResolvedValue(mockTypes);
        TransconnectApi.getUser.mockResolvedValue(mockUser);
        TransconnectApi.getUserPosts.mockResolvedValue(mockResources);
        TransconnectApi.deleteResource.mockResolvedValue({});
    });

    it("renders without crashing", () => {
        render(
            <MemoryRouter initialEntries={["/resources"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources" element={<ResourceList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <MemoryRouter initialEntries={["/resources"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources" element={<ResourceList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("fetches and displays resources", async () => {
        render(
            <MemoryRouter initialEntries={["/resources"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources" element={<ResourceList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            const resourceCards = screen.getAllByTestId("resource-card");
            expect(resourceCards.length).toBe(mockResources.length);
            expect(resourceCards[0]).toHaveTextContent(mockResources[0].name);
            expect(resourceCards[1]).toHaveTextContent(mockResources[1].name);
        });
    });

    it("displays message when no resources are found", async () => {
        TransconnectApi.getResources.mockResolvedValue([]);

        render(
            <MemoryRouter initialEntries={["/resources"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources" element={<ResourceList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByText("Sorry, no results were found!")).to.exist;
        });
    });

    it("handles resource deletion", async () => {
        render(
            <MemoryRouter initialEntries={["/resources"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources" element={<ResourceList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByTestId("resource-card")).to.exist;
        });

        fireEvent.click(screen.getByTestId("resource-card"));

        await waitFor(() => {
            expect(TransconnectApi.deleteResource).toHaveBeenCalledWith(101);
            expect(screen.queryByTestId("resource-card")).to.not.exist;
        });
    });

    it("filters resources based on search term", async () => {
        const searchTerm = "Resource One";
        render(
            <MemoryRouter initialEntries={["/resources"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources" element={<ResourceList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );


        fireEvent.change(screen.getByLabelText(/Enter search term.../i), { target: { value: searchTerm } });

        fireEvent.click(screen.getByText("Search"));

        await waitFor(() => {
            const resourceCards = screen.getAllByTestId("resource-card");
            expect(resourceCards.length).toBe(1);
            expect(resourceCards[0]).toHaveTextContent(searchTerm);
        });
    });

    it("filters resources based on selected type", async () => {
        render(
            <MemoryRouter initialEntries={["/resources"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources" element={<ResourceList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        fireEvent.click(screen.getByText("Type1"));

        await waitFor(() => {
            const resourceCards = screen.getAllByTestId("resource-card");
            expect(resourceCards.length).toBe(1);
            expect(resourceCards[0]).toHaveTextContent("Resource One");
        });
    });
});
