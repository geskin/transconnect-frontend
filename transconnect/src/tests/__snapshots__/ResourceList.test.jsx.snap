// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceList component > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root css-n5o2oh"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-ugkrwo-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r4:"
          id=":r4:-label"
        >
          Enter search term...
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r4:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-81qg8w"
            >
              <span>
                Enter search term...
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Search
      </button>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
        value=""
      >
        X
      </button>
    </div>
    <div>
      <p
        class="lead white-letters"
      >
        Sorry, no results were found!
      </p>
    </div>
    <div
      class="MuiBox-root css-16j30i3"
    >
      <a
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
        data-discover="true"
        href="/resources/new"
        tabindex="0"
      >
        Submit a resource
      </a>
    </div>
  </div>
</DocumentFragment>
`;
